designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+Through function call operator overloading, the callback writer class
+methods provide the appropriate formatting for different kinds of data.
+When the user sends a request to the interface to do some kind of inference given a model and data,
+the interfaces first instantiate the Stan model object and the writer callbacks,
+then pass these in to `stan::services` functions which invoke the appropriate algorithm.
+The use of writer callbacks keeps the inference algorithms output-format agnostic.
+The current `stan::services` layer utilities are completely Stan CSV format-centric and will need to be refactored.
+
+
+### Stan program outputs
+
+The Stan model class `write_array` method returns a vector of doubles for all
+parameter, transformed parameter, and generated quantities (non-local) variables,
+in order of declaration in the Stan program.
+
+The model class methods `constrained_param_names`, `unconstrained_param_names`, `get_constrained_sizedtypes`, `get_unconstrained_sizedtypes`
+provide name, type, and shape information.
+The reason that the model reports on both the constrained and unconstrained parameters is the Stan parameter types
+simplex, correlation, covariance, and Cholesky factors have different shapes on the constrained, unconstrained scale.
+For example, a simplex parameter is a vector whose elements sum to one on the constrained scale which means that it has $N-1$ free parameters
+when computing on the unconstrained scale.
+The Stan model parameter inputs and outputs are on the *constrained* scale, therefore the number of parameter values reported by the `write_array` method
+is reported by the `get_constrained_sizedtypes` method.
+The sampler computes on the *unconstrained* scale, therefore the size of the metric (inverse mass matrix) is reported by the `get_unconstrained_sizedtypes` method.
+(Note: A further complication for optimization is that the optimization algorithm computes on the unconstrained scale, which means that the Hessian of the covariance matrix is on the unconstrained scale.)
+
+
+### Inference algorithm outputs
+
+The outputs from the inference algorithm necessarily differ according to the method, and to a lesser degree, according to the developers.
+The NUTS-HMC sampler does full Bayesian inference by assembling a set of draws from the posterior distribution using an HMC sampler
+with tuning parameters step-size and metric.
+The `stan::services::util::mcmc_writer` class is constructed with a callback writer and provides methods
+`write_sample_names`, `write_sample_params`, `write_adapt_finish`, `write_timing`.
+
+- The `write_sample_names` and `write_sample_params` produce the CSV header row and data rows, respectively.
+The initial columns consist of sampler state and diagnostic information from the `stan::mcmc::sample` object,
+(called `sample_params`), the remaining columns are the Stan model variable names and values returned by
+the `constrained_param_names` and `write_array` methods.
```

Comment:
Is `constrained_param_names` a method or static class function?  If I recall, there are two forms, one that prints out just the names, and that's the same for constrained and unconstrained, and also one that prints out the whole thing as strings with indexes.

Reply:
`get_param_names` (sic) - names of model variables
`unconstrained_param_names`, `constrained_param_names` - names plus indices